monai/data/meta_tensor.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import warnings
+from typing import Callable
+
+import torch
+
+from monai.data.meta_obj import MetaObj, get_track_meta, get_track_transforms
+from monai.utils.enums import PostFix
+
+__all__ = ["MetaTensor"]
+
+
+class MetaTensor(MetaObj, torch.Tensor):
+    """
+    Class that inherits from both `torch.Tensor` and `MetaObj`, adding support for meta
+    data.
+
+    Metadata is stored in the form of a dictionary. Nested, an affine matrix will be
+    stored. This should be in the form of `torch.Tensor`.
+
+    Behavior should be the same as `torch.Tensor` aside from the extended
+    meta functionality.
+
+    Copying of information:
+        * For `c = a + b`, then auxiliary data (e.g., metadata) will be copied from the
+        first instance of `MetaTensor`.
+
+    Example:
+        .. code-block:: python
+
+            import torch
+            from monai.data import MetaTensor
+
+            t = torch.tensor([1,2,3])
+            affine = torch.eye(4)
+            meta = {"some": "info"}
+            m = MetaTensor(t, affine=affine, meta=meta)
+            m2 = m+m
+            assert isinstance(m2, MetaTensor)
+            assert m2.meta == meta
+
+    Notes:
+        - Older versions of pytorch (<=1.8), `torch.jit.trace(net, im)` may
+            not work if `im` is of type `MetaTensor`. This can be resolved with
+            `torch.jit.trace(net, im.as_tensor())`.
+        - For older versions of pytorch (<=1.7), `torch.save(m, fname); m=torch.load(fname)`
+            may return a `torch.Tensor` instead of MetaTensor`.
+        - A warning will be raised if in the constructor `affine` is not `None` and
+            `meta` already contains the key `affine`.
+    """
+
+    @staticmethod
+    def __new__(cls, x, affine: torch.Tensor | None = None, meta: dict | None = None, *args, **kwargs) -> MetaTensor:
+        return torch.as_tensor(x, *args, **kwargs).as_subclass(cls)  # type: ignore
+
+    def __init__(self, x, affine: torch.Tensor | None = None, meta: dict | None = None) -> None:
```

Comment:
I may be wrong but I think we could implement that using only `__new__` ?

Reply:
yup! 👍